React/testing.md

```diff
@@ -1 +1,299 @@
+
+**Question:** Can you do TDD with a Snapshot test?
+
+### Creating our First Snapshot
+
+Lets create a snapshot test for one of our untested components, `NewPetForm`.
+
+First create `src/NewPetForm.test.js`
+
+```javascript
+// src/NewPetForm.test.js
+import React from 'react';
+import NewPetForm from './NewPetForm';
+import { mount } from 'enzyme';
+
+describe('NewPetForm', () => {
+  test('that it matches an existing snapshot', () => {
+    const wrapper = mount( <NewPetForm addPetCallback={() => {} } />);
+
+    expect(wrapper).toMatchSnapshot();
+    wrapper.unmount();
+  });
+});
+```
+Now lets break the test.  Modify the `NewPetForm` component by adding a class attribute to one of the elements.  You should see something like this:
+
+![Failed Snapshot](./images/failedSnapShot.png)
+
+This is called a shapshot diff.  Notice the added content are highlighted in red with a `+` sign.  You can update the snapshot with the `u` key.  Update the snapshot now.
+
+**Question:** After updating the snapshot take out the class name.  What does the snapshot diff look like now?
+
+**Exercise** Create a test file and a test for the `Pet` component.
+
+You can see a solution on the testing branch of the repository.
+
+In the above example we used Enzyme's `mount` function to create the snapshot.  You can read about `mount` in Enzyme's documentation [here](https://airbnb.io/enzyme/docs/api/mount.html).  Using `mount` renders a component including any subcomponents fully in a virtual DOM.  Because it mounts the component in the DOM, it is important to call `.unmount` at the end of every test to prevent one test from affecting another.
+
+### Snapshots Files
+
+You may wonder, "Where are these snapshots saved?"  You can see the snapshot file in your components folder.
+
+![Snashot folder](./images/snapshotFolder.png)
+
+**Question** What does the snapshot file look like?
+
+Notice that the snapshot file looks a lot like the JSX rendered by the component, but it's a generated JSON file making it easier to compare to new snapshots.  You generally won't need to look at these files directly, but looking at the snapshot file should help you understand how Jest compares the current snapshot to the rendered content from the test.
+
+
+### Creating a Shallow Snapshot with enzyme
+
+This type of snapshot can work just fine for small components without subcomponents.  However with a complicated container component with dozens of subcomponents the snapshot can be thousands of lines long.  It would also be very fragile.  Any time one subcomponent changed, the snapshot would fail with an enormous diff.  A test that fails too often and is hard to parse gets ignored becoming useless.  **Snapshot test files should as a rule be as small as practical and specific to the component being tested**.
+
+To keep snapshot files smaller we can engage in **shallow rendering**.
+
+### Shallow Rendering
```

Comment:
I like that you do the regular `mount` and the `shallow` to show the difference. Can we just run the big snapshot first before discussing why we'd want the `shallow` one?